### DISCUSSION TITLE
cy.wrap() change command behaviour

### DISCUSSION THREAD

**Discussion by ValeronAI** [cypress-io/cypress#28234] [Questions and Help]
In case custom commands with previously yielded subject (`prevSubject: true`) should be used `cy.wrap()` function.

But two similars commands at the first glance in some case will give different results. For example:

```
cy.get("[data-cypress-id=test-element-wrapper]")
  .find("[data-cypress-id=test-element]")
  .should("have.length", 2);
```

and
```
cy.get("[data-cypress-id=test-element-wrapper]")
  .findByCypressId("test-element")
  .should("have.length", 2);
```
Where `findByCypressId` is defined:
```
Cypress.Commands.add(
    "findByCypressId",
    { prevSubject: true },
    (subject, selector, ...args) => {
        return cy.wrap(subject).find(`[data-cypress-id=${selector}]`, ...args);
    }
);
```

I expect that results of this commands will be similar. But in case DOM changing while test running, results will be different.

I have create simple example in which results of two above commands isn't similar. I test page on which change elements count after 1.5s after page initialising. Builtin `find` command has expected behaviour, it retries command for a few seconds and test successfully finished. But in case custom `findByCypressId` don't retry elements find so test failing. Below I'm leaving Vue single file component example and cypress test, also link to simple project example.

```
<script setup>
  import { ref } from "vue";

  const showFirstElement = ref(true);

  setTimeout(() => (showFirstElement.value = false), 1500);
</script>

<template>
  <ul>
    <li v-if="showFirstElement" data-cypress-id="test-element-wrapper">
      <span data-cypress-id="test-element">test item</span>
    </li>
    <li data-cypress-id="test-element-wrapper">
      <span data-cypress-id="test-element">test item</span>
    </li>
    <li data-cypress-id="test-element-wrapper">
      <span data-cypress-id="test-element">test item</span>
    </li>
  </ul>
</template>
```

Page tests:
```
describe("Wrap testing", () => {
    it("Use builtin find command", () => {
        cy.visit("/");

        cy.get("[data-cypress-id=test-element-wrapper]")
			.find("[data-cypress-id=test-element]")
			.should("have.length", 2);
    });
    it("Use custom command findByCypressId", () => {
        cy.visit("/");

        cy.get("[data-cypress-id=test-element-wrapper]")
			.findByCypressId("test-element")
			.should("have.length", 2);
    });
});
```

For testing please clone and run tests in public project: https://github.com/ValeronAI/cypress-test.

**Comment by portyu9** [comment]
Hi @ValeronAI 

Your reproduction is demonstrating the difference between a Cypress **query** and a normal custom **command**.

Built-in `.find()` is a query. Queries are linked together and Cypress can re-run that query chain when the downstream assertion retries.

Your custom command introduces a command boundary here:

```js
return cy.wrap(subject).find(...)
```

so it does not have the same retry semantics as adding another query to the original chain.

Since `findByCypressId()` is conceptually a query, define it with `Cypress.Commands.addQuery()` instead:

```js
Cypress.Commands.addQuery('findByCypressId', function (selector) {
  return (subject) => {
    return Cypress.$(subject).find(
      `[data-cypress-id="${selector}"]`
    )
  }
})
```

Then your original test can remain:

```js
cy.get('[data-cypress-id="test-element-wrapper"]')
  .findByCypressId('test-element')
  .should('have.length', 2)
```

The inner function returned by `addQuery()` is synchronous and Cypress may invoke it repeatedly while retrying the linked query chain. Do not put `cy.wrap()` or other Cypress commands inside that inner function.

Also note the `function` syntax above rather than an arrow function. Cypress's custom-query API uses `this` internally for query behavior such as timeouts.

`Cypress.Commands.addQuery()` was added in Cypress 12 specifically for this kind of retriable custom querying behavior.

Please mark this answer as accepted if it helped, thank you.
